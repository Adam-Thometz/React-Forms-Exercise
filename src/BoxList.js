import React, {useState} from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
  const [boxes, setBoxes] = useState([])

  const addBox = ({id, color, width, height}) => {
    setBoxes(boxes => [...boxes, {id, color, width, height}])
  }
  const removeBox = id => {
    setBoxes(boxes => boxes.filter(b => b.id !== id))
  }

  return (
    <div className="BoxList">
      <NewBoxForm addBox={addBox} />
      {boxes.map(b => 
        <Box key={b.id} id={b.id} width={b.width} height={b.height} color={b.color} handleRemove={removeBox} />)}
    </div>
  )
}

export default BoxList