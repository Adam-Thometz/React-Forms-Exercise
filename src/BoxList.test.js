import { render, fireEvent, waitFor } from '@testing-library/react';
import BoxList from './BoxList';

const addBox = async (boxList, color = "blue", height = "100", width = "200") => {
  const colorInput = boxList.getByLabelText("Color")
  const heightInput = boxList.getByLabelText("Height")
  const widthInput = boxList.getByLabelText("Width")
  const submitBtn = boxList.getByText("Create Box")
  
  await waitFor(() => {
    fireEvent.change(colorInput, {target: {value: color}})
    fireEvent.change(heightInput, {target: {value: height}})
    fireEvent.change(widthInput, {target: {value: width}})
    fireEvent.click(submitBtn)
  })
}

it('renders without crashing', () => {
  render(<BoxList />);
});

it('matches snapshot', () => {
  const {asFragment} = render(<BoxList />)
  expect(asFragment()).toMatchSnapshot()
});

it("can add a new box", async () => {
  const boxList = render(<BoxList />)
  expect(boxList.queryByText("Remove")).not.toBeInTheDocument()

  await addBox(boxList)

  const removeBtn = boxList.getByText("Remove")
  expect(removeBtn).toBeInTheDocument()
  expect(removeBtn.previousSibling).toHaveStyle(`
    backgroundColor: blue;
    height: 100px;
    width: 200px;
  `)
})

it("can remove a box", async () => {
  const boxList = render(<BoxList />)
  await addBox(boxList)

  const removeBtn = boxList.getByText("Remove")
  fireEvent.click(removeBtn)
  expect(removeBtn).not.toBeInTheDocument()
})