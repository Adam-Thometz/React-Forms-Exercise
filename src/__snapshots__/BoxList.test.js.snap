// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        placeholder="Enter color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="width in px"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        placeholder="height in px"
        type="text"
        value=""
      />
      <button>
        Create Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
